File: chapter4/nodejs/com/dotComIt/learnWith/services/TaskService.js
/**
 * Created by jhouser on 6/4/2016.
 */

var dateFormatter = require("dateformat");
var databaseConnection = require("./../database/DatabaseConnection");
var responseHandler = require("./../server/JSONResponseHandler");

function getFilteredTasks(response, queryString) {
    var resultObject = {};
    var callback = '';
    if(queryString.callback != undefined){
        callback = queryString.callback;
    }

    if((queryString.filter == undefined)){
        resultObject.error = 1;
        responseHandler.execute(response, resultObject, callback);
    } else {
        var json = JSON.parse( queryString.filter );
        var firstOne = true;

        var query = "select tasks.*, taskCategories.taskCategory from tasks left outer join taskCategories on (tasks.taskCategoryID = taskCategories.taskCategoryID) ";

        if((json.taskCategoryID != undefined) && (json.taskCategoryID != "0")){
            if(firstOne){
                query = query + "Where "
                firstOne = false;
            } else {
                query = query + "and "
            }
            query = query + " taskCategories.taskCategoryID = " +
                json.taskCategoryID + " ";
        }

        if(json.completed != undefined){
            if(firstOne){
                query = query + "Where "
                firstOne = false;
            } else {
                query = query + "and "
            }
            if(json.completed == true){
                query = query + " completed = 1 ";
            } else {
                query = query + " completed = 0 ";
            }
        }

        if(json.startDate != undefined){
            if(firstOne){
                query = query + "Where "
                firstOne = false;
            } else {
                query = query + "and "
            }
            query = query + " dateCreated >= '" + json.startDate + "' ";
        }
        if(json.endDate != undefined){
            if(firstOne){
                query = query + "Where "
                firstOne = false;
            } else {
                query = query + "and "
            }
            query = query + " dateCreated <= '" + json.endDate + "' ";
        }

        if(json.scheduledStartDate != undefined){
            if(firstOne){
                query = query + "Where "
                firstOne = false;
            } else {
                query = query + "and "
            }
            query = query + " dateScheduled >= '" + json.scheduledStartDate + "' ";
        }
        if(json.scheduledEndDate != undefined){
            if(firstOne){
                query = query + "Where "
                firstOne = false;
            } else {
                query = query + "and "
            }
            query = query + " dateScheduled <= '" + json.scheduledEndDate + "' ";
        }


        query = query + " order by dateCreated ";

        console.log(query);

        var dataQuery = databaseConnection.executeQuery(query,
            function(result) {
                // Process Query Results here
                console.log(result);
                resultObject.error = 0;
                for (var x=0;x < result.recordset.length ; x++ ){
                    result.recordset[x].dateCreated =dateFormatter(result.recordset[x].dateCreated, "mm/dd/yyyy");
                    if(result.recordset[x].dateCompleted){
                        result.recordset[x].dateCompleted =dateFormatter(result.recordset[x].dateCompleted, "mm/dd/yyyy");
                    } else {
                        result.recordset[x].dateCompleted = '';
                    }
                    if(result.recordset[x].dateScheduled){
                        result.recordset[x].dateScheduled.setDate(result.recordset[x].dateScheduled.getDate() + 1);
                        result.recordset[x].dateScheduled =dateFormatter(result.recordset[x].dateScheduled, "mm/dd/yyyy");
                    } else {
                        result.recordset[x].dateScheduled = '';
                    }
                }

                resultObject.resultObject = result.recordset;
                responseHandler.execute(response, resultObject, callback);

            },
            function(err){
                // process error results here
                console.log(err);
                console.log('close get filtered tasks')
                resultObject.error = 1;
                responseHandler.execute(response, resultObject, callback);
            }
        )


    }

}

function getTaskCategories(response, queryString) {
    var resultObject = {};
    var callback = '';
    if(queryString.callback != undefined){
        callback = queryString.callback;
    }
    var query = "select * from taskCategories order by taskCategory";

    var dataQuery = databaseConnection.executeQuery(query,
        function(result) {
            // query result function here
            result.recordset.unshift({taskCategoryID:0,taskCategory:"All Categories"});
            resultObject.error = 0;
            resultObject.resultObject = result.recordset;
            responseHandler.execute(response, resultObject, callback);
        },
        function(err) {
            // query error function here
            console.log(err);
            console.log('close get task categories')
            resultObject.error = 1;
            responseHandler.execute(response, resultObject, callback);
            return;
        }
    )

}


exports.getFilteredTasks = getFilteredTasks;
exports.getTaskCategories = getTaskCategories;

File: chapter4/java/src/main/java/com/dotComIt/learnWith/services/TaskService.java
package com.dotComIt.learnWith.services;

import com.dotComIt.learnWith.vos.ResultObjectVO;
import com.dotComIt.learnWith.vos.TaskCategoryVO;
import com.dotComIt.learnWith.vos.TaskFilterVO;
import com.dotComIt.learnWith.vos.TaskVO;

import java.sql.Connection;
import java.sql.Date;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.util.ArrayList;

public class TaskService {

    Connection connection = null;

    public TaskService(Connection connection) {
        this.connection = connection;
    }

    public ResultObjectVO getFilteredTasks(TaskFilterVO filter){

        String SQL;
        PreparedStatement pstmt = null;
        ResultSet rs = null;
        ResultObjectVO ro = new ResultObjectVO();

        try {

            // using a prepared statement which is better suited against injection attacks
            SQL = "select tasks.*, taskCategories.taskCategory from tasks join taskCategories on (tasks.taskCategoryID = taskCategories.taskCategoryID)";
            SQL += " where 0=0 ";
            if(filter.getCompleted() != null){
                SQL += "and completed = ? ";
            }
            if(filter.getTaskCategoryID() != 0){
                SQL += "and tasks.taskCategoryID = ? ";
            }
            if(filter.getStartDate() != null){
                SQL += "and dateCreated >= ? ";
            }
            if(filter.getEndDate() != null){
                SQL += "and dateCreated <= ? ";
            }
            if(filter.getScheduledStartDate() != null){
                SQL += "and dateScheduled >= ? ";
            }
            if(filter.getScheduledEndDate() != null){
                SQL += "and dateScheduled <= ? ";
            }
            SQL += "order by dateCreated ";

            System.out.println(SQL );
            pstmt = connection.prepareStatement(SQL);
            int parameterCounter = 1;
            if(filter.getCompleted() != null){
                pstmt.setBoolean(parameterCounter, filter.getCompleted());
                parameterCounter++;
            }
            if(filter.getTaskCategoryID() != 0){
                pstmt.setInt(parameterCounter, filter.getTaskCategoryID());
                parameterCounter++;
            }
            if(filter.getStartDate() != null){
                pstmt.setDate(parameterCounter,  Date.valueOf( filter.getStartDate() ));
                parameterCounter++;
            }
            if(filter.getEndDate() != null){
                pstmt.setDate(parameterCounter,  Date.valueOf( filter.getEndDate() ));
                parameterCounter++;
            }
            if(filter.getScheduledStartDate() != null){
                pstmt.setDate(parameterCounter,  Date.valueOf( filter.getScheduledStartDate() ));
                parameterCounter++;
            }
            if(filter.getScheduledEndDate() != null){
                pstmt.setDate(parameterCounter,  Date.valueOf( filter.getScheduledEndDate() ));
                parameterCounter++;
            }

            rs = pstmt.executeQuery();

            ro.setError(false);

            ArrayList<TaskVO> tasks = new ArrayList<TaskVO>();
            while(rs.next()){

                // save results
                TaskVO task = new TaskVO();
                task.setTaskID(rs.getInt("taskID"));
                task.setTaskCategoryID(rs.getInt("taskCategoryID"));
                task.setTaskCategory(rs.getString("taskCategory"));
                task.setUserID(rs.getInt("userID"));
                task.setDescription(rs.getString("description"));
                // this could be wrong
                if(rs.getBoolean("completed")){
                    task.setCompleted(true);
                } else {
                    task.setCompleted(false);
                }
                // dates
                task.setDateCreated(rs.getDate("dateCreated").toLocalDate());

                if(rs.getDate("dateCompleted") != null){
                    task.setDateCompleted(rs.getDate("dateCompleted").toLocalDate());
                }
                if(rs.getDate("dateScheduled") != null){
                    task.setDateScheduled(rs.getDate("dateScheduled").toLocalDate());
                }

                tasks.add(task);
            }
            ro.setResultObject(tasks);
        }
        catch (Exception e) {
            System.out.println("TaskService getFilteredTasks Exception");
            e.printStackTrace();
            ro.setError(true);
        }
        finally {
            if (rs != null) try { rs.close(); } catch(Exception e) {}
            if (pstmt != null) try { pstmt.close(); } catch(Exception e) {}
            if (connection != null) try { connection.close(); } catch(Exception e) {}
        };

        return ro;
    };


    public ResultObjectVO getTaskCategories(){

        String SQL;
        PreparedStatement pstmt = null;
        ResultSet rs = null;
        ResultObjectVO ro = new ResultObjectVO();

        try {

            // using a prepared statement which is better suited against injection attacks
            SQL = "select * from taskCategories order by taskCategory";
            pstmt = connection.prepareStatement(SQL);
            rs = pstmt.executeQuery();

            ro.setError(false);

            ArrayList<TaskCategoryVO> taskCategories = new ArrayList<TaskCategoryVO>();
            while(rs.next()){

                // save results
                TaskCategoryVO taskCategory = new TaskCategoryVO();
                taskCategory.setTaskCategoryID(rs.getInt("taskCategoryID"));
                taskCategory.setTaskCategory(rs.getString("taskCategory"));

                taskCategories.add(taskCategory);
            }
            ro.setResultObject(taskCategories);
        }
        catch (Exception e) {
            System.out.println("TaskService getTaskCategories Exception");
            e.printStackTrace();
            ro.setError(true);
        }
        finally {
            if (rs != null) try { rs.close(); } catch(Exception e) {}
            if (pstmt != null) try { pstmt.close(); } catch(Exception e) {}
            if (connection != null) try { connection.close(); } catch(Exception e) {}
        };

        return ro;
    };

}
